# Drug Efficacy Dataset for Asthma: Data Understanding and Initial Exploration

## Project Overview

In this project, I am working with a drug efficacy dataset related to asthma patients. The main outcome of interest is `post_index_exacerbations365`, which represents the number of asthma exacerbations during the first year after starting therapy.

The goal of this project is to build an end-to-end machine learning workflow that can help predict asthma-related outcomes using pre-index clinical and claims-based patient information. Before building any model, I first need to understand the dataset carefully, check the structure of the variables, identify possible data quality problems, and think about what kind of modeling approach may be appropriate.

This notebook focuses on the first stage of the project: data loading, initial inspection, and basic understanding of the dataset.


## Initial Questions Before Modeling

Before starting machine learning modeling, I want to answer a few basic questions about the dataset:

1. How many rows and columns are present in the dataset?
2. What are the input features and what is the target variable?
3. Are there missing values or duplicate records?
4. What are the data types of the variables?
5. Are the variables mostly binary, continuous, or count-based?
6. What does the outcome variable `post_index_exacerbations365` look like?
7. Should this problem be treated as a regression problem, a count prediction problem, or possibly a classification problem?

These questions will help me decide the next steps for exploratory data analysis and modeling.

In [1]:
# Basic libraries for data handling and exploration
import pandas as pd
import numpy as np

# Display settings to make notebook output easier to read
pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 100)
pd.set_option("display.float_format", "{:.4f}".format)

## Loading the Dataset

The dataset was downloaded from Kaggle and saved inside the `data/raw/` folder. I am keeping the original raw data unchanged so that the project remains reproducible. Any cleaned or processed versions of the data will be saved later in the `data/processed/` folder.

In [2]:
# Define the relative file path
data_path = "../data/raw/DrugEfficacyDataset.csv"

# Read the dataset
df = pd.read_csv(data_path)

# Display the first few rows
df.head()

,patid,index_age,previous_asthma_drugs,total_pre_index_cannisters_365,post_index_exacerbations365,pneumonia,sinusitis,acute_bronchitis,acute_laryngitis,upper_respiratory_infection,gerd,rhinitis,adherence,total_pre_index_charge,pre_asthma_days,pre_asthma_charge,pre_asthma_pharma_charge,drug_s,female,log_charges,log_asthma_charge
0,1073754155,14,1,1,0,0,0,1,0,1,0,0,0.0845,1224.7675,1,314.0000,218.1300,0,0,7.1105,5.3851
1,1073799394,21,1,2,2,0,1,1,0,0,0,1,0.0845,20290.5343,0,0.0000,44.9800,0,0,9.9179,3.8062
2,1073854918,62,1,0,0,0,0,0,0,0,0,0,0.7384,2964.2542,4,480.0000,99.2600,0,0,7.9944,4.5977
3,1073898249,30,1,2,0,0,0,1,0,1,0,0,0.0845,3223.7088,0,0.0000,59.5800,0,1,8.0783,4.0873
4,1073913003,40,1,1,0,0,1,0,1,1,0,1,0.5068,1287.2544,2,689.0000,29.7900,0,0,7.1603,3.3942


## First Look at the Dataset

After loading the dataset, I will first check the shape of the data and preview the first few rows. This gives me an initial idea of how many patient records and variables are available, and whether the columns were read correctly.

In [3]:
# Check number of rows and columns
df.shape

(18215, 21)

In [4]:
# Display all column names
df.columns.tolist()

['patid',
 'index_age',
 'previous_asthma_drugs',
 'total_pre_index_cannisters_365',
 'post_index_exacerbations365',
 'pneumonia',
 'sinusitis',
 'acute_bronchitis',
 'acute_laryngitis',
 'upper_respiratory_infection',
 'gerd',
 'rhinitis',
 'adherence',
 'total_pre_index_charge',
 'pre_asthma_days',
 'pre_asthma_charge',
 'pre_asthma_pharma_charge',
 'drug_s',
 'female',
 'log_charges',
 'log_asthma_charge']

In [5]:
# Check data types and non-null counts
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 18215 entries, 0 to 18214
Data columns (total 21 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   patid                           18215 non-null  int64  
 1   index_age                       18215 non-null  int64  
 2   previous_asthma_drugs           18215 non-null  int64  
 3   total_pre_index_cannisters_365  18215 non-null  int64  
 4   post_index_exacerbations365     18215 non-null  int64  
 5   pneumonia                       18215 non-null  int64  
 6   sinusitis                       18215 non-null  int64  
 7   acute_bronchitis                18215 non-null  int64  
 8   acute_laryngitis                18215 non-null  int64  
 9   upper_respiratory_infection     18215 non-null  int64  
 10  gerd                            18215 non-null  int64  
 11  rhinitis                        18215 non-null  int64  
 12  adherence                       18215 non-n

## Initial Dataset Structure Observation

The dataset contains **18,215 patient records** and **21 columns**. Each row appears to represent one asthma patient, and the columns include demographic information, prior asthma-related drug/treatment history, diagnosis indicators, healthcare charge variables, medication adherence, drug group assignment, and the post-index outcome.

The target variable is:

- `post_index_exacerbations365`

This variable represents the number of asthma exacerbations during the year after initiating therapy. Since the values are non-negative counts, this may not be a simple continuous regression problem. It may require thinking about count-based modeling, regression modeling, or converting the outcome into a binary classification problem later.

At this stage, I will not decide the final modeling approach yet. First, I need to better understand the variable types, data quality, and outcome distribution.

In [6]:
# To display a few random rows to get a broader sense of the data
df.sample(5, random_state=22)

,patid,index_age,previous_asthma_drugs,total_pre_index_cannisters_365,post_index_exacerbations365,pneumonia,sinusitis,acute_bronchitis,acute_laryngitis,upper_respiratory_infection,gerd,rhinitis,adherence,total_pre_index_charge,pre_asthma_days,pre_asthma_charge,pre_asthma_pharma_charge,drug_s,female,log_charges,log_asthma_charge
16113,1215780903,53,1,0,0,0,0,0,0,0,0,1,0.3760,1433.0595,1,227.0000,845.7000,1,0,7.2676,6.7402
1585,1125859796,33,1,1,0,0,1,1,0,1,0,1,0.0845,1427.5297,3,510.0000,54.3900,0,1,7.2637,3.9962
12535,1505760435,44,1,0,0,0,1,0,0,0,0,1,0.2534,8722.6079,4,885.0000,173.3000,0,0,9.0737,5.1550
4609,1202872464,47,1,0,0,0,1,1,0,0,0,0,0.0845,26599.7906,0,0.0000,98.4000,0,1,10.1887,4.5890
566,1092017818,48,1,1,0,0,0,0,0,0,0,0,0.5068,8027.5014,1,70.0000,22.4800,0,1,8.9906,3.1126


In [7]:
# To check summary statistics for numerical variables
df.describe().T

,count,mean,std,min,25%,50%,75%,max
patid,18215.0000,1317072654.2669,154570755.9664,1073754155.0000,1183442347.5000,1282388763.0000,1493698756.5000,1771898675.0000
index_age,18215.0000,38.3046,15.2182,12.0000,26.0000,41.0000,50.0000,65.0000
previous_asthma_drugs,18215.0000,1.0000,0.0000,1.0000,1.0000,1.0000,1.0000,1.0000
total_pre_index_cannisters_365,18215.0000,0.7937,0.7235,0.0000,0.0000,1.0000,1.0000,2.0000
post_index_exacerbations365,18215.0000,0.1746,0.6333,0.0000,0.0000,0.0000,0.0000,14.0000
pneumonia,18215.0000,0.0502,0.2184,0.0000,0.0000,0.0000,0.0000,1.0000
sinusitis,18215.0000,0.3197,0.4664,0.0000,0.0000,0.0000,1.0000,1.0000
acute_bronchitis,18215.0000,0.2605,0.4389,0.0000,0.0000,0.0000,1.0000,1.0000
acute_laryngitis,18215.0000,0.0169,0.1287,0.0000,0.0000,0.0000,0.0000,1.0000
upper_respiratory_infection,18215.0000,0.2217,0.4154,0.0000,0.0000,0.0000,0.0000,1.0000


## Important First Observations from `df.info()` and `df.describe()`

From the initial inspection, I observed the following:

1. There are **no missing values** based on the non-null counts.
2. All columns are numeric, which is convenient for machine learning modeling.
3. The column `patid` appears to be a patient identifier and should not be used as a predictive feature.
4. The column `previous_asthma_drugs` has the same value for every patient, so it may not provide useful predictive information.
5. Several variables are binary indicator variables, such as `pneumonia`, `sinusitis`, `acute_bronchitis`, `female`, and `drug_s`.
6. The outcome variable `post_index_exacerbations365` has a mean of about **0.17**, a median of **0**, and a maximum of **14**. This suggests that most patients had no exacerbation after treatment, while a small number had multiple exacerbations.
7. The charge-related variables are highly skewed. For example, `total_pre_index_charge` has a median around **3,155** but a maximum above **875,000**. This suggests the presence of extreme high-cost patients.
8. The dataset already includes log-transformed charge variables: `log_charges` and `log_asthma_charge`, which may be more useful for modeling than the original skewed charge variables.

In [8]:
# To check missing values in each column
missing_values = df.isnull().sum().sort_values(ascending=False)

missing_values

patid                             0
index_age                         0
previous_asthma_drugs             0
total_pre_index_cannisters_365    0
post_index_exacerbations365       0
pneumonia                         0
sinusitis                         0
acute_bronchitis                  0
acute_laryngitis                  0
upper_respiratory_infection       0
gerd                              0
rhinitis                          0
adherence                         0
total_pre_index_charge            0
pre_asthma_days                   0
pre_asthma_charge                 0
pre_asthma_pharma_charge          0
drug_s                            0
female                            0
log_charges                       0
log_asthma_charge                 0
dtype: int64

In [9]:
# Check missing value percentage
missing_percent = (df.isnull().mean() * 100).sort_values(ascending=False)

missing_summary = pd.DataFrame({
    "missing_count": missing_values,
    "missing_percent": missing_percent
})

missing_summary

,missing_count,missing_percent
patid,0,0.0000
index_age,0,0.0000
previous_asthma_drugs,0,0.0000
total_pre_index_cannisters_365,0,0.0000
post_index_exacerbations365,0,0.0000
pneumonia,0,0.0000
sinusitis,0,0.0000
acute_bronchitis,0,0.0000
acute_laryngitis,0,0.0000
upper_respiratory_infection,0,0.0000


## Missing Value Observation

Based on the initial missing value check, the dataset does not appear to have missing values. This is helpful because I may not need imputation at this stage.

However, I should still be careful because some values may be coded as zero even when they represent absence of service use or no recorded charge. For example, `pre_asthma_charge` has a minimum value of 0, which may be meaningful rather than missing.

In [10]:
# To check number of duplicated rows
duplicate_count = df.duplicated().sum()

duplicate_count

np.int64(0)

In [11]:
# Lets check whether patient IDs are unique
unique_patients = df["patid"].nunique()
total_rows = df.shape[0]

print("Total rows:", total_rows)
print("Unique patient IDs:", unique_patients)
print("Duplicate patient IDs:", total_rows - unique_patients)

Total rows: 18215
Unique patient IDs: 18215
Duplicate patient IDs: 0


## Duplicate Record Observation

The dataset contains **0 fully duplicated rows**.

The dataset contains **18215 unique patient IDs** out of **18,215 total rows**. This check is important because `patid` appears to be a patient identifier. If each row represents one patient, then the number of unique patient IDs should match the number of rows.

If duplicate patient IDs are present, I will need to investigate whether they represent repeated records, multiple treatment episodes, or possible data duplication.

In [12]:
# Check number of unique values in each column
unique_counts = df.nunique().sort_values()

unique_counts

previous_asthma_drugs                 1
acute_bronchitis                      2
sinusitis                             2
pneumonia                             2
rhinitis                              2
gerd                                  2
upper_respiratory_infection           2
acute_laryngitis                      2
female                                2
drug_s                                2
total_pre_index_cannisters_365        3
post_index_exacerbations365          14
pre_asthma_days                      46
index_age                            54
adherence                           353
pre_asthma_charge                  3946
pre_asthma_pharma_charge           8375
log_asthma_charge                  8375
log_charges                       17968
total_pre_index_charge            17968
patid                             18215
dtype: int64

## Constant or Low-Information Variables

I checked the number of unique values in each column to identify variables that may not be useful for modeling.

One important observation from the summary statistics is that `previous_asthma_drugs` has only one value for all patients. Since it does not vary across patients, it cannot help a machine learning model distinguish between different outcomes. Therefore, this variable may be removed before modeling.

The `patid` column is also not a clinical predictor. It is an identifier and should not be used as a model input because it does not represent meaningful patient characteristics.

In [13]:
# Grouping Variables by Type
# Define target and ID columns
target_col = "post_index_exacerbations365"
id_col = "patid"

# Binary indicator variables based on dataset description and summary statistics
binary_cols = [
    "pneumonia",
    "sinusitis",
    "acute_bronchitis",
    "acute_laryngitis",
    "upper_respiratory_infection",
    "gerd",
    "rhinitis",
    "drug_s",
    "female"
]

# Count / discrete numeric variables
count_cols = [
    "index_age",
    "previous_asthma_drugs",
    "total_pre_index_cannisters_365",
    "pre_asthma_days"
]

# Continuous charge and adherence variables
continuous_cols = [
    "adherence",
    "total_pre_index_charge",
    "pre_asthma_charge",
    "pre_asthma_pharma_charge",
    "log_charges",
    "log_asthma_charge"
]

print("Target column:", target_col)
print("ID column:", id_col)
print("Binary columns:", binary_cols)
print("Count/discrete columns:", count_cols)
print("Continuous columns:", continuous_cols)

Target column: post_index_exacerbations365
ID column: patid
Binary columns: ['pneumonia', 'sinusitis', 'acute_bronchitis', 'acute_laryngitis', 'upper_respiratory_infection', 'gerd', 'rhinitis', 'drug_s', 'female']
Count/discrete columns: ['index_age', 'previous_asthma_drugs', 'total_pre_index_cannisters_365', 'pre_asthma_days']
Continuous columns: ['adherence', 'total_pre_index_charge', 'pre_asthma_charge', 'pre_asthma_pharma_charge', 'log_charges', 'log_asthma_charge']


## Feature Grouping

To understand the dataset more clearly, I grouped the variables into different types:

- **Identifier variable:** `patid`
- **Target variable:** `post_index_exacerbations365`
- **Binary diagnosis or demographic variables:** pneumonia, sinusitis, bronchitis, laryngitis, respiratory infection, GERD, rhinitis, drug group, and sex
- **Count or discrete variables:** age, previous asthma drugs, SABA canister count, and pre-asthma treatment days
- **Continuous variables:** adherence and healthcare charge-related variables

This grouping will help during exploratory data analysis because binary, count, and continuous variables need different types of summaries and visualizations.

In [14]:
# Summary of the target variable
df[target_col].describe()

count   18215.0000
mean        0.1746
std         0.6333
min         0.0000
25%         0.0000
50%         0.0000
75%         0.0000
max        14.0000
Name: post_index_exacerbations365, dtype: float64

In [15]:
# Frequency of each exacerbation count
target_counts = df[target_col].value_counts().sort_index()

target_counts

post_index_exacerbations365
0     16124
1      1525
2       333
3       115
4        48
5        32
6        12
7         9
8         9
9         3
10        2
12        1
13        1
14        1
Name: count, dtype: int64

In [16]:
# Frequency and percentage of each exacerbation count
target_distribution = pd.DataFrame({
    "count": df[target_col].value_counts().sort_index(),
    "percent": df[target_col].value_counts(normalize=True).sort_index() * 100
})

target_distribution

,count,percent
post_index_exacerbations365,,
0,16124,88.5205
1,1525,8.3722
2,333,1.8282
3,115,0.6313
4,48,0.2635
5,32,0.1757
6,12,0.0659
7,9,0.0494
8,9,0.0494


## Initial Target Variable Observation

The outcome variable `post_index_exacerbations365` is a count variable. The minimum value is **0** and the maximum value is **14**.

The mean is about **0.17**, while the median is **0**. This suggests that the outcome is highly concentrated at zero. In other words, most patients did not experience an exacerbation during the post-index year.

This is an important modeling issue. If most patients have zero exacerbations, a model may perform well by predicting low values for everyone, but it may still fail to identify patients who actually have exacerbations. Because of this, I may need to evaluate the problem in more than one way:

1. Predicting the exact number of exacerbations as a regression/count problem.
2. Predicting whether a patient had any exacerbation as a binary classification problem.
3. Comparing standard machine learning models with count-aware approaches.

In [17]:
# Create a binary version of the target:
# 0 = no exacerbation
# 1 = at least one exacerbation
df["had_exacerbation"] = (df[target_col] > 0).astype(int)

df["had_exacerbation"].value_counts()

had_exacerbation
0    16124
1     2091
Name: count, dtype: int64

In [18]:
# Percentage of patients with and without exacerbation
df["had_exacerbation"].value_counts(normalize=True) * 100

had_exacerbation
0   88.5205
1   11.4795
Name: proportion, dtype: float64

## Binary Outcome Idea

Although the original outcome is a count variable, I created a temporary binary outcome called `had_exacerbation` for exploration:

- `0`: No exacerbation during the post-index year
- `1`: At least one exacerbation during the post-index year

This does not replace the original target variable. It is only created to help understand the data from a clinical risk perspective. Later, I can decide whether the project should focus on count prediction, binary classification, or both.

Asthma drug efficacy modeling